Ignore rseq syscall on riscv to fix SE workloads built with newer glibc

jiegec opened the pull request:
The rseq syscall is invoked by newer glibc on e.g. Debian Trixie. It
leads to syscall out of range failure in SE mode running riscv binaries.

Reproducible via:

```
pushd tests/test-progs/hello
riscv64-linux-gnu-gcc -static src/hello.c -o bin/riscv/linux/hello
popd
./build/RISCV/gem5.opt configs/learning_gem5/part1/simple-riscv.py
```

Fails with:

```
fatal: Syscall 293 out of range
```

We can safely ignore 293 to make it work.

Additionally, to make pre-commit hook happy, run clang-format in a separate commit.

Fixes https://github.com/gem5/gem5/issues/2391.

jiegec committed:
arch-riscv: Return -ENOSYS for rseq syscall in riscv64

The rseq syscall is invoked by newer glibc on e.g. Debian Trixie. It
leads to syscall out of range failure in SE mode running riscv binaries.

Reproducible via:

pushd tests/test-progs/hello
riscv64-linux-gnu-gcc -static src/hello.c -o bin/riscv/linux/hello
popd
./build/RISCV/gem5.opt configs/learning_gem5/part1/simple-riscv.py

Fails with:

fatal: Syscall 293 out of range

We can safely use ignoreWithEnosysFunc for 293 to fix it.

Fixes issue #2391.

Signed-off-by: Jiajie Chen <c@jia.je>

powerjg reviewed (commented):
From https://lwn.net/Articles/883104/ it sounds like glibc is using this to initialize threads now. This change may work for hello_world, but I worry that it's going to cause issues with any multithreaded applications. From the comments in that article, a better solution would be to return `ENOSYS`.

jiegec committed:
sim: Add ignoreWithEnosysFunc to syscall emulation

The existing ignoreFunc always returns success, but in some cases we
want to return -ENOSYS to signify that the syscall is missing
explicitly. The new ignoreWithEnosysFunc is added for this purpose.

Signed-off-by: Jiajie Chen <c@jia.je>

jiegec commented:
> From https://lwn.net/Articles/883104/ it sounds like glibc is using this to initialize threads now. This change may work for hello_world, but I worry that it's going to cause issues with any multithreaded applications. From the comments in that article, a better solution would be to return `ENOSYS`.

Added:

```
    sim: Add enosysFunc to syscall emulation

    The existing ignoreFunc always returns success, but in some cases we
    want to return -ENOSYS to signify that the syscall is missing
    explicitly. The new enosysFunc is added for this purpose.
```

And uses enosysFunc instead of ignoreFunc for syscall 293.

powerjg commented on src/sim/syscall_emul.hh:
```diff
@@ -141,6 +141,10 @@ SyscallReturn ignoreFunc(SyscallDesc *desc, ThreadContext *tc);
 
+/// Handler for unimplemented syscalls that return -ENOSYS to the target
+/// program.
+SyscallReturn enosysFunc(SyscallDesc *desc, ThreadContext *tc);
```
I like this! Could you rename to ignoreWithEnosys? This way it will be clear in the syscall list that the syscall isn't enosys but that we are ignoring it with enosys.

jiegec replied:
Done.

powerjg reviewed (commented):
Thanks for the update. I like this solution a lot. One small feedback, though.

erin-le commented:
@jiegec: Hello, it looks like the entirety of `se_workload.cc` has been reformatted with `clang-format`. In order to avoid mixing style changes with unrelated changes, the CI tests only check the modified/added lines in a PR using `git-clang-format`.
Could you drop the current reformatting commit and run `git-clang-format` on your changes? To do so, please run `util/run-git-clang-format.py --ci-pr-base-commit <commit hash of commit before the first commit of the PR>` from the gem5 directory. You can also look at `CI Tests / clang-format-check` for an example. Thank you!

jiegec commented:
> @jiegec: Hello, it looks like the entirety of `se_workload.cc` has been reformatted with `clang-format`. In order to avoid mixing style changes with unrelated changes, the CI tests only check the modified/added lines in a PR using `git-clang-format`. Could you drop the current reformatting commit and run `git-clang-format` on your changes? To do so, please run `util/run-git-clang-format.py --ci-pr-base-commit <commit hash of commit before the first commit of the PR>` from the gem5 directory. You can also look at `CI Tests / clang-format-check` for an example. Thank you!

Done, however the reformatted range is extended to the whole syscallDescs64 definition by the tool. Not sure if it is intended, though.

powerjg commented:
> Done, however the reformatted range is extended to the whole syscallDescs64 definition by the tool. Not sure if it is intended, though.

I think that's a reasonable compromise.

powerjg reviewed (approved):
Thanks for the updates and apologies for the clang-format annoyance. We're going through some growing pains with it.
